src/thunderbots/software/ai/navigator/path_planner/astar.h

```diff
@@ -0,0 +1,149 @@
+#pragma once
+
+// I apologize if your CLion slows to a crawl
+#include <boost/graph/astar_search.hpp>
+#include <boost/graph/grid_graph.hpp>
+
+#include "ai/world/field.h"
+#include "path_planner.h"
+namespace AStar
+{
+    // TODO: potentially move a lot of stuff out of this header
+
+    // number of graph vertices per metre
+    // TODO: make this a dynamic parameter or passed in the constructor
+    // to AStarGridGraph
+    static constexpr unsigned long GRID_VERTEX_DENSITY = 20;
+    // edge cost type
+    typedef double edge_cost_t;
+    // 2D grid graph type
+    typedef boost::grid_graph<2> GridGraph2D;
+    // vertex type - a 2D point on a grid
+    typedef GridGraph2D::vertex_descriptor GridVertex;
+
+    // a hash function for a grid vertex, necessary for all the maps
+    // that boost::astar_search uses
+    struct grid_vertex_hash : std::unary_function<GridVertex, std::size_t>
+    {
+        std::size_t operator()(const GridVertex &gp) const;
+    };
+
+    class AStarGridGraph
+    {
+       public:
+        AStarGridGraph() = delete;
+
+        /**
+         * Constructs a grid graph from a Field object.
+         *
+         * @param field the field to create a graph for
+         */
+        explicit AStarGridGraph(const Field &field);
+
+        /**
+         * Converts a grid point to a Point on the field.
+         *
+         * @param grid_v the grid vertex to convert to a Point.
+         */
+        Point gridPointToPoint(const GridVertex &grid_v);
+
+        /**
+         * Finds the nearest vertex on the grid graph to the
+         * given point.
+         *
+         * @param point the point to find a grid vertex close to it
+         */
+        GridVertex nearestGridVertex(const Point &point);
+
+        /**
+         * Returns a const reference to the grid graph itself.
+         * @return the underlying grid graph.
+         */
+        const GridGraph2D &graph();
+
+        /**
+         * The distance between vertices on the grid graph.
+         * @return the distance between vertices on the grid graph.
+         */
+        constexpr double gridPointDistance();
+
+       private:
+        std::unique_ptr<GridGraph2D> field_graph;
+
+        // this map is used to convert from the grid points that A* will
+        // return, into Points that can be used for navigation
+        std::vector<std::pair<Point, GridVertex>> grid_points_list;
```

Comment:
is this not something we can get from the grid vertex?

Reply:
there's no way that I know of to do this with boost::grid_graph, because the constructor only takes the dimensions of the grid and nothing else, and the vertex_descriptor is fixed to be of type `boost::array<(number of dimensions), size_t, size_t, ...>`.